### ISSUE TITLE
Points saving and rendering problem

### ISSUE THREAD

**Issue by Czaki** [napari/napari#2288]
## 🐛 Bug

From one user I got a points csv file with points marked. I see that points on some layers are not showing. Then I check csv file and found that part of the points has such coordinates: 

```
116.0,0.0,12.000000000000002,104.44800634733548,171.98994564952477
117.0,0.0,13.000000000000002,95.2150207065153,32.176163088533485
```

The user works with data with noninteger scaling

There are two bugs at once. 

Points should be saved in proper image coordinates. So `12.0` not `12.000000000000002`

Points with coordinate `12.000000000000002` should be rendered on layer 12. 

## To Reproduce

1. Download attached zip
2. load tif and csv files to napari
3. notice missed points on layer 12 and 13

## Additional context

[Cy5_component2.zip](https://github.com/napari/napari/files/6005902/Cy5_component2.zip)

**Comment by jni** [comment]
@Czaki are you saying point 12.000000000000002 is *not* rendered on slice 12? That's *definitely* a bug.

As to the issue of floating point error that is giving rise to this... floating point is always tricky but I suppose it should be safe enough to round to, I dunno, 12 decimal places 😂 

Any thoughts @sofroniewn?

**Comment by Czaki** [comment]
> @Czaki are you saying point 12.000000000000002 is _not_ rendered on slice 12? That's _definitely_ a bug.

yes

**Comment by Czaki** [comment]
What did you think about use round function before saving in image coordinate space?

**Comment by jni** [comment]
I sort of implied that round would be sort of ok but to some ludicrous precision. I can see a plain-old-round being troublesome in cases where the slicing is *not* actually at integer coordinates. Note that this is not "image coordinate space", it's whatever results from the points layer scale and translate parameters. The points layer does not know whether this is image coordinates or some other space.

**Comment by sofroniewn** [comment]
> @Czaki are you saying point 12.000000000000002 is not rendered on slice 12? That's definitely a bug.

Agreed this sounds like a bug, I can take a look when I get a chance and report back here. Will be something funny going on in our Dims model https://github.com/napari/napari/blob/v0.4.5/napari/components/dims.py

**Comment by Czaki** [comment]
> I sort of implied that round would be sort of ok but to some ludicrous precision. I can see a plain-old-round being troublesome in cases where the slicing is _not_ actually at integer coordinates. Note that this is not "image coordinate space", it's whatever results from the points layer scale and translate parameters. The points layer does not know whether this is image coordinates or some other space.

No. The points are saved in image coordinate space. I already mentioned it earlier on zulip. 
The original data has a scale equal to `1, 300, 64, 64` reader form image metadata (and from division by ~300 error my comes)

**Comment by andy-sweet** [comment]
The main bug here was fixed by #2289.

I can't reproduce the other bug with the built-in CSV reader/writer. I wrote the following test to try, but it passes for me on macOS.

```python
def test_write_read_points_csv_with_whole_numbers(tmpdir):
    dtype = 'float'
    filename = os.path.join(tmpdir, 'test.csv')
    data = np.array([
        [116, 0, 12, 104, 171],
        [117, 0, 13, 95, 32],
    ], dtype=dtype)
    column_names = tuple(f'axis-{i}' for i in range(data.shape[1]))

    io.write_csv(filename, data, column_names=column_names)

    read_data, _, _ = io.read_csv(filename, require_type='any')
    read_data = np.array(read_data, dtype=dtype)

    np.testing.assert_array_equal(data, read_data)
```

@Czaki : are you able to provide the `Points.data` raw values before you wrote them to CSV? Or some operation that produces them?

I would argue 12.000000000000002 is as valid as 12 as a point coordinate even in the points data space - I understand they may well be feature points associated with integer pixel indices of an image, but that is not always the case.

Agreed that choosing a format string with some fixed precision (likely something like `{:.10g}`) could be a good idea, but it won't solve the problem in general, so I'm leaning towards closing this issue.

**Comment by Czaki** [comment]
Here is data produced a few minutes ago with napari 0.4.15 by clicking on the layer. There are two arrays inside. `coordinates` with raw points and scale with layer `scale` parameter  
[data.zip](https://github.com/napari/napari/files/8401355/data.zip)

And the head of csv file:

```
index,axis-0,axis-1,axis-2,axis-3
0.0,0.0,11.999999999999998,77.5551566465027,32.32186273534832
1.0,0.0,11.999999999999998,54.07232525135826,60.719705352732305
2.0,0.0,11.999999999999998,79.46654989959585,68.91139072313152
3.0,0.0,11.999999999999998,66.9059656649837,96.49006480347556
4.0,0.0,13.999999999999998,91.20796559716806,60.9927615317456
5.0,0.0,13.999999999999998,35.50450507845334,83.3833682108368
6.0,0.0,13.999999999999998,47.51897695503888,56.07775030950606
7.0,0.0,13.0,108.68356105401975,62.084986247798845
```

[test.csv](https://github.com/napari/napari/files/8401363/test.csv)

**Comment by andy-sweet** [comment]
Oh I think I get it now. The `11.999999999999998` comes from a slider position. Given that slider positions in general correspond to world coordinates that might also be unavoidable. But for a simple case like this I might hope to avoid it.

Thanks for the data! By inspecting it in ipython, I think we can see that we lose the precision in `Points.data` not when writing the CSV.

```python
In [1]: a = np.load('data.npz')
In [2]: a['coordinates'][0, 1]
Out[2]: 11.999999999999998
```

My guess is this might be related to extent/transforms, so I'll take a look and see if we might be able to avoid it in this case.

**Comment by andy-sweet** [comment]
I think I understand this now and it is related to dims/transforms. Consider the following failing test that just needs an image layer.

```python
def test_exact_point_coordinates_on_image_with_scale():
    """See https://github.com/napari/napari/issues/2288"""
    np.random.seed(0)
    # Add a 3D image with and odd shape and anisotropic scale.
    data = np.random.rand(9, 7, 5) 
    scale = (210, 70, 70)
    viewer = ViewerModel()
    image = viewer.add_image(data, scale=scale)
    
    # Move the current slice to an odd position in the first dimension.
    target_data_point = (5,) + viewer.dims.current_step[1:]
    viewer.dims.current_step = target_data_point
    
    # Convert the current point back to the image data space.
    current_world_point = viewer.dims.point
    current_data_point = image.world_to_data(current_world_point)

    # The data point does not exactly match the original target due to lost precision
    # when inverting the transform or when applying it.
    np.testing.assert_array_equal(current_data_point, target_data_point)
```

When using a points layer with similar dimensions/scale, clicking on a points coordinate calls `_points_mouse_bindings.add` which then calls `Layer.world_to_data` in a similar way to the above.

Given that the dims model accommodates multiple layers (with different transform parameter values like scale) using a shared world coordinate space, this issue seems generally unavoidable.

One bigger/general fix I can think of is to add something like snap-to-grid like functionality in napari, so that we could round coordinates for the added point to the nearest point in its data grid (e.g. that might overlap with the center of a pixel in an image that the points layer is annotating). We'd need to check all places that this would apply, and also work out how to expose this graphically. If we want a solution like that, I suggest creating a fresh issue to capture the goals of that work and closing this issue, as the main bug here has been fixed.

Also open to other ideas. If anyone can think of any quick fixes that don't generally break the dims world coordinate system, then I'd be more open to applying such a fix to close this issue.

**Comment by andy-sweet** [comment]
@Czaki : do you have any other ideas for quick fixes that could resolve the remaining issue? And what do you think about a more general solution like the snap-to-grid option I mentioned above?

**Comment by Czaki** [comment]
Looking that every, reported by me, problem is around 15 digits after dot then maybe around to 1e-5 or 1e-10 could be onough?

It is similar to your gród solution but left more flexibility to users?

From the other hand integer coordinates also looks promising.

**Comment by andy-sweet** [comment]
Ah I see, so maybe as quick fix, add a non-required parameter to `napari.plugins._builtins.napari_write_points` for the float precision/format and default it to something like 10 digits after the decimal point (i.e. '%.10f')? Similar to `pandas.DataFrame.to_csv`'s [`float_format`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html)?

I don't think there is an easy way to expose that as a parameter in the GUI (i.e. it will always take the default), but maybe it's OK to just expose it in the API for now?

The snap-to-grid option would be more useful if you actually wanted to `Points.data` to actually store integers (at least when those can be represented by floats), in which case we could round when applying the inverse transform on the mouse click. So it's a more general fix in that respect, but definitely requires much more work.
